### PR TITLE
Bg/curve signature wrapping

### PR DESCRIPTION
**Pull request by SimpleApp**
- Expose curv_sign & curv_verify functions as swift's methods in SignalModule
- Change the target language version to swift 5

### CODE REVIEW

File: libsignal-protocol-swift/Misc/Signature.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    // privatekey: Taken from KeyPair
+    public static func sign(payload: Data, withPrivateKey privateKey: Data) -> Result<Data, SignatureError> {
+
+        // Convert private key to [UInt8]
+        let privateBuffer = privateKey.signalBuffer
+        defer { signal_buffer_free(privateBuffer) }
+        let privLength = signal_buffer_len(privateBuffer)
+        let privData = signal_buffer_data(privateBuffer)!
```
Comment (thibauddavid) [inline libsignal-protocol-swift/Misc/Signature.swift]: force unwrap ?

File: libsignal-protocol-swift/Misc/Signature.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            curve_decode_private_point($0, privData, privLength, Signal.context)
+        }
+        guard keyDecodeResult == 0 else {
+            return .failure(.keyDecodeError(SignalError(value: keyDecodeResult)))
+        }
+        defer { ec_private_key_destroy(privKey) }
+
+        var signatureBuf: OpaquePointer? = nil
+        let signResult = withUnsafeMutablePointer(to: &signatureBuf) { signatureBufPtr in
+            return payload.withUnsafeBytes {
```
Comment (thibauddavid) [inline libsignal-protocol-swift/Misc/Signature.swift]: Isn't `withUnsafeBytes` deprecated ? You should load its inner `UnsafeRawBufferPointer` as the type you want, see https://stackoverflow.com/a/55378571/2219009

File: libsignal-protocol-swift/Misc/Signature.swift [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    // publicKey : taken from KeyPair
+    public static func verify(signature: Data, for payload: Data, withPublicKey publicKey: Data) -> Result<Bool, SignatureError> {
+
+        // Convert public key
+        let publicBuffer = publicKey.signalBuffer
+        defer { signal_buffer_free(publicBuffer) }
+        let pubLength = signal_buffer_len(publicBuffer)
+        let pubData = signal_buffer_data(publicBuffer)!
```
Comment (thibauddavid) [inline libsignal-protocol-swift/Misc/Signature.swift]: Force unwrap

File: libsignal-protocol-swift/Misc/Signature.swift [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            curve_decode_point($0, pubData, pubLength, Signal.context)
+        }
+
+        guard keyDecodeResult == 0 else {
+            return .failure(.keyDecodeError(SignalError(value: keyDecodeResult)))
+        }
+        defer { ec_public_key_destroy(pubKey) }
+
+        let verifyRes = signature.withUnsafeBytes { signaturePtr in
+            payload.withUnsafeBytes { payloadPtr in
```
Comment (thibauddavid) [inline libsignal-protocol-swift/Misc/Signature.swift]: Deprecated ?

**Review by thibauddavid — CHANGES_REQUESTED** [review]

File: libsignal-protocol-swift/Misc/Signature.swift (hunk 3 above)
Comment (SimpleApp) [reply libsignal-protocol-swift/Misc/Signature.swift]: good catch. The other parts of the wrapper uses this pattern a lot though... We may want to change that everywhere in a future PR

File: libsignal-protocol-swift/Misc/Signature.swift (hunk 4 above)
Comment (SimpleApp) [reply libsignal-protocol-swift/Misc/Signature.swift]: The project was using swift 4.2 so it raised no deprecation warning. I moved it to swift 5, but this pattern is used everywhere. 

Do you think it's better to use the same pattern everywhere, or use the new one here and keep the old one in the other parts ?

File: libsignal-protocol-swift/Misc/Signature.swift (hunk 2 above)
Comment (SimpleApp) [reply libsignal-protocol-swift/Misc/Signature.swift]: It wasn't in swift4.2 (the actual target for the project). I switch to 5 and fixed it. it was actually simpler than in the stackoverflow. in the case of byte arrays you don't even need to load (or even rebind the memory). By default it's already in the good units and there are no memory alignment issues.

**Review by thibauddavid — APPROVED** [review]

**Pull request merged**